src/main/scala/org/allenai/plugins/DockerBuildPlugin.scala

```diff
@@ -112,23 +117,175 @@ object DockerBuildPlugin extends AutoPlugin {
+        // A `clean` will wipe these out if needed.
+        if (!destinationFile.exists || destinationFile.getName.contains("-SNAPSHOT")) {
+          IO.copyFile(file, destinationFile)
+        }
+        destinationFile
+    }
+
+    // Remove any items in `lib` that are stale.
+    val staleItems = lib.listFiles.toSet -- copiedFiles
+    staleItems.foreach(_.delete())
+
+    imageDirectory
+  }
+
+  /** Task to stage the main docker image for the project. */
+  lazy val mainImageStageDef: Def.Initialize[Task[File]] = Def.task {
+    val logger = Keys.streams.value.log
+
+    val dockerfile = dockerfileLocation.value
+    if (!dockerfile.exists) {
+      sys.error(s"No Dockerfile found at $dockerfile .\n" +
+        "Maybe you should generate one with the `generateDockerfile` task?")
+    }
+
+    logger.info("Staging main image ...")
+
+    // Create the destination directory.
+    val imageDirectory = mainImageDir.value
+    IO.createDirectory(imageDirectory)
+
+    // Create the destination for libraries.
+    val lib = new File(imageDirectory, "lib")
+    if (!lib.exists) {
+      IO.createDirectory(lib)
+    }
+
+    // Copy the Dockerfile.
+    IO.copyFile(dockerfile, new File(imageDirectory, "Dockerfile"))
+    // Copy the mappings.
+    dockerCopyMappings.value.foreach {
+      case (maybeRelativeSource, relativeDestination) =>
+        // Make any relative path relative to the source directory.
+        val source = if (maybeRelativeSource.isAbsolute) {
+          maybeRelativeSource
+        } else {
+          new File(sourceMain.value, maybeRelativeSource.toString)
+        }
+        if (source.exists) {
+          val destination = new File(imageDirectory, relativeDestination)
+          if (source.isDirectory) {
+            IO.createDirectory(destination)
+            IO.copyDirectory(source, destination)
+          } else {
+            IO.copyFile(source, destination)
+          }
+        }
+    }
+
+    // Copy the local project jars.
+    HelperDefs.localDependencies.value.foreach {
+      case (file, destination) =>
+        val destinationFile = new File(lib, destination)
+        IO.copyFile(file, destinationFile)
+    }
+
+    imageDirectory
   }
 
   /** Adds the settings to configure the `dockerBuild` command. */
   override def projectSettings: Seq[Def.Setting[_]] = Seq(
-    dockerfileLocation := Keys.sourceDirectory.value.toPath.resolve("docker/Dockerfile").toFile,
+    dockerfileLocation := {
+      new File(sourceMain.value, "docker" + File.separatorChar + "Dockerfile")
+    },
     dockerCopyMappings := defaultCopyMappings.value,
     dockerImageRegistryHost := AI2_PRIVATE_REGISTRY,
     dockerImageNamePrefix := Keys.organization.value.stripPrefix("org.allenai."),
     dockerImageName := Keys.name.value,
     dockerImageBase := DEFAULT_BASE_IMAGE,
-    dockerWorkdir := "/local/deploy/" + dockerImageName.value
+    dockerWorkdir := dockerImageName.value,
```

Comment:
I think this default should be simply `"/stage"`, not image dependent.  We can leave the override in case someone wants to combine images in unexpected ways, but I don't want to encourage that pattern up front.